azure/services/virtualmachines/virtualmachines.go

```diff
@@ -137,6 +137,17 @@ func (s *Service) Reconcile(ctx context.Context) error {
 			return errors.Wrap(err, "failed to generate OS Profile")
 		}
 
+		var generation compute.HyperVGenerationType
+		vmImage, err := s.Scope.GetVMImage()
+		if err != nil {
+			return errors.Wrap(err, "failed to get VM Image")
+		}
+		if vmImage.HyperVGeneration == "" {
```

Comment:
What if someone sets `vmImage.HyperVGeneration == "V1"`? Should this be set in `s.Scope.GetVMImage()`? If it is, probably could get rid of the conditional and just have `compute.HyperVGenerationType(string(vmImage.HyperVGeneration))`.

Reply:
if its set to `"v1"` the `if` condition would still be fine, basically, I'm setting it to v1 as a default since we don't publish "v2" images yet, eventually we should use v2.

As for moving this to the `scope`; usually translating the types to the `azure sdk` types that will be used in the call to create happens just before that call, so trying to be consistent.

Reply:
> As for moving this to the scope; usually translating the types to the azure sdk types that will be used in the call to create happens just before that call, so trying to be consistent.

`s.generateOSProfile` calls `s.GetVMImage` to build the `*compute.StorageProfile`.

Reply:
the func `generateOSProfile` is part of virtualmachine.go. not scope.

I might be mis understanding, I though you want me to move the if condition to set the `generation` var to the scope.GetVMImage, so was just saying scope doesn't use anything from the `compute` types

Reply:
By setting the `vmImage.HyperVGeneration = "V1" or "V2"` in `s.GetVMImage`, you can avoid compute types, encapsulate the behavior in `s.GetVMImage`, and have simpler code building the compute VM structure with a single line: 
```go
InstanceView: &compute.VirtualMachineInstanceView{
		HyperVGeneration: compute.HyperVGenerationType(string(vmImage.HyperVGeneration)),
},
```

Reply:
My fault. I see what you were saying about `generateOSProfile`. Maybe `generateOSProfile` returns the `vmImage.HyperVGeneration` too.

Reply:
ok, I see what you are saying, will update the code

Reply:
thanks, your suggestions make the code better

Reply:
changed it, not sure if this is exactly what you had in mind, PTAL

Reply:
wdyt about the structure below?

```go
// generateStorageProfile generates a pointer to a compute.StorageProfile which can utilized for VM creation.
func (s *Service) generateStorageProfile(ctx context.Context, vmSpec azure.VMSpec, sku resourceskus.SKU) (*compute.StorageProfile, HyperVGenerationType, error) {
        //....
	return storageProfile, vmImage.HyperVGeneration, nil
}
```

```go
osProfile, hyperVGeneration, err := s.generateOSProfile(ctx, vmSpec)
if err != nil {
     return errors.Wrap(err, "failed to generate OS Profile")
}

// ... omitted code //
      InstanceView: &amp;compute.VirtualMachineInstanceView{
            HyperVGeneration: compute.HyperVGenerationType(string(hyperVGeneration)),
      },
```